Dcikatic/sol 487 filter

dino-cikatic opened the pull request:
This PR represents the work necessary to enable courseware search to support searching of cohorted content. The first instance of courseware search did not support cohorted content; this represents the changes necessary to deliver such. (see https://openedx.atlassian.net/browse/SOL-495)

edx-webhook commented:
Thanks for the pull request, @dino-cikatic! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: http://openedx-webhooks.herokuapp.com/github/process_pr?number=7640&repo=edx%2Fedx-platform

dino-cikatic commented:
@martynjames please ignore the test changes for now since I'm still trying to fix those and am aware they are failing.

martynjames commented on common/lib/xmodule/xmodule/modulestore/courseware_index.py:
```diff
@@ -64,13 +64,21 @@ def _fetch_item(item_location):
             return item
 
-        def index_item_location(item_location, current_start_date):
+        def index_item_location(item_location, current_start_date, content_groups=None):
```
You'll want to integrate in the same fashion into the latest changes for indexing.

martynjames commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -137,7 +137,17 @@ def reindex_course_and_check_access(course_key, user):
         raise PermissionDenied()
-    return CoursewareSearchIndexer.do_course_reindex(modulestore(), course_key)
+
+    course = modulestore().get_course(course_key)
```
This feels like it should really be part of the indexing job, not part of the view. Also, with the changes for SOL-174, we are loading the full course as part of the indexing, so it will be much more efficient to do it therein.

martynjames commented on common/lib/xmodule/xmodule/modulestore/courseware_index.py:
```diff
@@ -83,11 +91,13 @@ def index_item_location(item_location, current_start_date):
             if item.has_children:
                 for child_loc in item.children:
-                    index_item_location(child_loc, current_start_date)
+                    if getattr(item, 'content_groups', None):
```
Seems like you just pass in the value of the call to getattr; the function already accommodates when it is None., it will certainly read quite a bit better

martynjames commented on lms/lib/courseware_search/lms_filter_generator.py:
```diff
@@ -0,0 +1,49 @@
+
+    def filter_dictionary(self, **kwargs):
+        """ base implementation which filters via start_date """
+        filter_dictionary = {"start_date": DateRange(None, datetime.utcnow())}
```
I think it would be better to call the base class implementation here to get the start date involved - that way, if it ever changes, then the changes will propagate

martynjames commented on common/test/acceptance/tests/lms/test_lms_cohorted_courseware_search.py:
```diff
@@ -0,0 +1,245 @@
+
+
+@attr('shard_1')
+class CoursewareSearchCohortTest(ContainerBase):
```
It seems like we'd really want a normal python unit test for this - see the newly added `cms/djangoapps/contentstore/tests/test_courseware_index.py` file in master - might be possible to add some tests therein, and not worry about needing to do bok choy for this specific test.

bokchoy is necessary to test the indexing => searching flow within the environment, we already have this - it might be easier to have just unit tests for cohort search.

On the other hand these seem fine, but I'd like to have cohort tests in the regular unit tests, so you can choose to keep these or not, but I'd really like unit tests.

dino-cikatic replied:
Since these work now, I would like to keep them (no harm in some extra testing). I'll add some more unit tests.

martynjames replied:
Super!

martynjames commented:
Lots of pending changes due to some of the significant changes for SOL-174. I think you'll need to rebase (feature/dashboard-search has been rebased) in order to get latest, and then a lot more of my comments will seem clearer. They almost all revolve around updating the code herein to work with the changes made for SOL-174.

edx-webhook commented:
Thanks for the pull request, @dino-cikatic! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: http://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=7640

martynjames commented:
@dino-cikatic - please address the failing tests herein

dino-cikatic commented:
jenkins test this please

marjev commented:
test this please

dino-cikatic commented:
jenkins test this please

dino-cikatic commented:
jenkins test this please

marjev commented on cms/djangoapps/contentstore/courseware_index.py:
```diff
@@ -150,7 +162,13 @@ def index_item(item, skip_index=False):
                 for child_item in item.get_children():
-                    index_item(child_item, skip_index=skip_child_index)
+                    temp_content = getattr(item, 'content_groups', None)
+                    if temp_content and temp_content is not []:
```
I think `temp_content is not []` should be replaced with `temp_content != []` if you wish to check that `temp_content` doesn't have an empty list for its value.

marjev commented on cms/djangoapps/contentstore/courseware_index.py:
```diff
@@ -163,7 +181,9 @@ def index_item(item, skip_index=False):
-
+                item_index['content_groups'] = getattr(item, 'content_groups', None)
+                if item_index['content_groups'] == []:
+                    item_index['content_groups'] = None
```
I think this could be in one line:

`item_index['content_groups'] = None if item_index['content_groups'] == []`

dino-cikatic replied:
That won't work (no trailing ifs).

martynjames replied:
I don't think this line is even necessary - `[]` is falsy in python, so it shouldn't matter if it is `[]` or `None` if a simple `if` statement is used.

martynjames replied:
I think that we can only add `content_groups` to the index where necessary:
```
if item_content_groups:
    item_index['content_groups'] = item_content_groups
```

dino-cikatic replied:
I've actually set it to None because these values get stored differently in a different searcher backend. Searching with this value set to None will return everything that has this field undefined/None, but not ones that have content groups defined to empty array/list.

martynjames replied:
Ah, that explains it - thank you!

marjev commented on cms/djangoapps/contentstore/courseware_index.py:
```diff
@@ -163,7 +181,9 @@ def index_item(item, skip_index=False):
                 item_index['id'] = item_id
                 if item.start:
                     item_index['start_date'] = item.start
```
not your doing but this can be placed in one line:

`item_index['start_date'] = item.start if item.start`

dino-cikatic replied:
That won't work (no trailing ifs).

marjev commented on cms/djangoapps/contentstore/courseware_index.py:
```diff
@@ -174,8 +194,16 @@ def index_item(item, skip_index=False):
+                                module['usage_key_string'] = kwargs['usage_key_string']
                 for item in structure.get_children():
-                    index_item(item)
+                    index_item(item, False, groups_usage_info)
```
Perhaps omitting the 2nd argument in this call, and naming the 3rd one, since the 2nd parameter is set to `False` by default.

marjev commented on cms/djangoapps/contentstore/courseware_index.py:
```diff
@@ -174,8 +194,16 @@ def index_item(item, skip_index=False):
+                if cls.DOCUMENT_TYPE == "courseware_content":
+                    groups_usage_info = GroupConfiguration.get_content_groups_usage_info(modulestore, structure).items()
+                    if groups_usage_info:
+                        for name, group in groups_usage_info:  # pylint: disable=unused-variable
```
Should this be rewritten in the following fashion `for group in groups_usage_info.values():` because `name` is unused here. This way, we could omit the pylint comment as well. Perhaps @martynjames might give his opinion on this.

martynjames replied:
Seems like the right thing to do if it works!

dino-cikatic commented:
jenkins test this please

dsego commented on cms/djangoapps/contentstore/courseware_index.py:
```diff
@@ -150,7 +162,13 @@ def index_item(item, skip_index=False):
                 for child_item in item.get_children():
-                    index_item(child_item, skip_index=skip_child_index)
+                    temp_content = getattr(item, 'content_groups', None)
+                    if temp_content and temp_content != []:
```
Empty sequences are false. (https://www.python.org/dev/peps/pep-0008/).

dsego commented on cms/djangoapps/contentstore/courseware_index.py:
```diff
@@ -150,7 +162,13 @@ def index_item(item, skip_index=False):
                     (triggered_at is not None and (triggered_at - item.subtree_edited_on) > reindex_age)
                 for child_item in item.get_children():
-                    index_item(child_item, skip_index=skip_child_index)
+                    temp_content = getattr(item, 'content_groups', None)
```
Might want to rename to something more meaningful, i.e. item_content_groups.

dsego commented on common/test/acceptance/tests/lms/test_lms_cohorted_courseware_search.py:
```diff
@@ -0,0 +1,245 @@
+        """
+        Reindex course content on studio course page
+        """
+
```
Extra newline?

dsego commented on lms/lib/courseware_search/lms_filter_generator.py:
```diff
@@ -0,0 +1,46 @@
+                partition_group = CourseUserGroupPartitionGroup.objects.get(course_user_group=cohort)
+            except CourseUserGroupPartitionGroup.DoesNotExist:
+                pass
+            if partition_group:
```
You could use an else clause instead.

martynjames commented on cms/djangoapps/contentstore/courseware_index.py:
```diff
@@ -174,8 +194,16 @@ def index_item(item, skip_index=False):
         try:
             with modulestore.branch_setting(ModuleStoreEnum.RevisionOption.published_only):
                 structure = cls._fetch_top_level(modulestore, structure_key)
+                groups_usage_info = None
```
This is now adding operations that I believe are specific to course indexing to the base class for both library and course indexing.

Some of it will need to remain, and it is unlikely that there will be content_groups for libraries and so most of this will be skipped. However, I feel that making the logic dependent upon the DOCUMENT_TYPE is a little fragile.

I think I'd prefer to have a base (empty) implementation, say
```
@classmethod
def fetch_group_usage(cls, modulestore, structure):
    return None
```

then the override for CoursewareSearchIndexer looks like:
```
@classmethod
def fetch_group_usage(cls, modulestore, structure):
    groups_usage_info = GroupConfiguration.get_content_groups_usage_info(modulestore, structure).items()
    if groups_usage_info:
        for group in groups_usage_info.values():
            for module in group:
                view, args, kwargs = resolve(module['url'])  # pylint: disable=unused-variable
                module['usage_key_string'] = kwargs['usage_key_string']
    return groups_usage_info
```

martynjames commented on cms/djangoapps/contentstore/courseware_index.py:
```diff
@@ -174,8 +194,16 @@ def index_item(item, skip_index=False):
+                                module['usage_key_string'] = kwargs['usage_key_string']
                 for item in structure.get_children():
-                    index_item(item)
+                    index_item(item, groups_usage_info=groups_usage_info)
```
... following on from above this line becomes:
```
                    index_item(
                        item,
                        groups_usage_info=cls.fetch_group_usage(modulestore, structure)
                    )
```

martynjames commented on cms/djangoapps/contentstore/courseware_index.py:
```diff
@@ -143,6 +145,16 @@ def index_item(item, skip_index=False):
+                item.content_groups = []
+                for name, group in groups_usage_info:
+                    for module in group:
+                        if str(item.location) == str(module['usage_key_string']):
```
use `unicode` instead of `str` here, please

martynjames commented on cms/djangoapps/contentstore/courseware_index.py:
```diff
@@ -143,6 +145,16 @@ def index_item(item, skip_index=False):
+                        if str(item.location) == str(module['usage_key_string']):
+                            item.content_groups.append(name)
+
+            if content_groups and not getattr(item, 'content_groups', None):
```
It's weird that we're calling `getattr(item, 'content_groups', None)` here, and then below on line 165...

martynjames commented on cms/djangoapps/contentstore/courseware_index.py:
```diff
@@ -150,7 +162,13 @@ def index_item(item, skip_index=False):
-                    index_item(child_item, skip_index=skip_child_index)
+                    item_content_groups = getattr(item, 'content_groups', None)
+                    if item_content_groups:
+                        index_item(child_item, skip_index=skip_child_index,
```
Nitpick / Optional: once the line is long enough, I prefer to place each argument on it's own line like this:
```
    index_item(
        child_item,
        skip_index=skip_child_index,
        groups_usage_info=groups_usage_info,
        content_groups=item.content_groups
    )
```

martynjames commented on cms/djangoapps/contentstore/tests/test_courseware_index.py:
```diff
@@ -29,6 +33,12 @@
 from contentstore.courseware_index import CoursewareSearchIndexer, LibrarySearchIndexer, SearchIndexingError
 from contentstore.signals import listen_for_course_publish, listen_for_library_update
 
+import json
```
I'd prefer to see the order of the imports in sensible grouping:
see https://github.com/edx/edx-platform/wiki/Python-Guidelines#imports-order

martynjames commented on cms/djangoapps/contentstore/courseware_index.py:
```diff
@@ -143,6 +145,16 @@ def index_item(item, skip_index=False):
 
+            if groups_usage_info:
+                item.content_groups = []
+                for name, group in groups_usage_info:
```
This seems like a strange double loop in here. It feels like we'd be better off twisting the data within groups_usage_info so that we can find matches for this item more efficiently.

Seems like we could build a dict with keys based upon `unicode(module['usage_key_string'])` that would return the array of associated content groups.. then lines 148-153 could become:
`item_content_groups = groups_usage_info.get(item.location, []) if groups_usage_info and else []`

and then we can always refer to item_content_groups without having to use `getattr...` so often

martynjames commented on cms/djangoapps/contentstore/courseware_index.py:
```diff
@@ -143,6 +145,16 @@ def index_item(item, skip_index=False):
                 return
 
+            if groups_usage_info:
+                item.content_groups = []
```
I'm not understanding why these need to be an injected property of the item, wouldn't a variable `item_content_groups` suffice herein?

martynjames commented on cms/djangoapps/contentstore/courseware_index.py:
```diff
@@ -143,6 +145,16 @@ def index_item(item, skip_index=False):
+                            item.content_groups.append(name)
+
+            if content_groups and not getattr(item, 'content_groups', None):
+                item.content_groups = content_groups
```
This seems like we're adding all content groups to the index for items that have no specification for content groups - I'm not sure I understand why this is necessary.

dino-cikatic replied:
There was a problem with indexing deleted content group permissions. If an item had content group and then it was removed I had to manually set it to None. If I didn't index stuff that had empty content group, I would not reindex these.

martynjames commented on cms/djangoapps/contentstore/courseware_index.py:
```diff
@@ -150,7 +162,13 @@ def index_item(item, skip_index=False):
                 for child_item in item.get_children():
-                    index_item(child_item, skip_index=skip_child_index)
+                    item_content_groups = getattr(item, 'content_groups', None)
+                    if item_content_groups:
```
Don't really need this conditional here - if we pass through `item_content_groups`, then the check for `if content_groups` will be falsy if `item_content_groups==[]`; so if we feed it on through to the child call, everything should just work :-)

dino-cikatic commented:
jenkins test this please

dino-cikatic commented:
jenkins test this please

martynjames commented on cms/djangoapps/contentstore/courseware_index.py:
```diff
@@ -163,7 +177,9 @@ def index_item(item, skip_index=False):
                     item_index['start_date'] = item.start
-
+                item_index['content_groups'] = item_content_groups
+                if item_index['content_groups'] == []:
```
Nit: This can now go in one line
`item_index['content_groups'] = item_content_groups if item_content_groups else None`

martynjames commented on cms/djangoapps/contentstore/tests/test_courseware_index.py:
```diff
@@ -783,3 +793,245 @@ def test_search_disabled(self, store_type):
         self._perform_test_using_store(store_type, self._test_exception)
+
+
+class GroupConfigurationSearchMongo(CourseTestCase, MixedWithOptionsTestCase):
```
If we're inheriting from the `MixedWithOptionsTestCase` then we have both stores defined within the test case anyway. Wondering why you're not following the same pattern (using ddt) as the other TestCases that inheirit from `MixedWithOptionsTestCase`

dino-cikatic replied:
I'm also inheriting CourseTestCase, which inherits ModuleStoreTestCase, which has store switch built-in using this approach. It looked simpler to me and requires less test code to be store-aware, but I can switch to ddt if that is what you prefer.

martynjames commented:
jenkins test this please

martynjames commented:
jenkins test this please

martynjames commented:
jenkins test this please

martynjames commented:
@cahrens - We've gone through a number of review passes here - finally ready for T&L to take a look at this one - Thanks!

martynjames commented:
@dino-cikatic - After the changes, I seem to have it not quite functioning correctly. I think that the cohort lookup needs to strip the version and branch information from the item.location like this:
`item_location = item.location.version_agnostic().replace(branch=None)`

Also, the contents of the `groups_usage_info` seems to only include the groups usage within the vertical (which is likely more than 1) with no entries for the individual usage therein, so I get both pieces of html content indexed with both cohort identifiers:
e.g.
```
         {
            "_index": "courseware_index",
            "_type": "courseware_content",
            "_id": "block-v1:TestX+CRT-101+2015_1+type@html+block@cb91b1d7ea2247c7a9b6992eb63b2c7e",
            "_score": 1,
            "_source": {
               "start_date": "2015-01-01T00:00:00+00:00",
               "content": {
                  "html_content": "This is content for Cohort B",
                  "display_name": "Cohort B"
               },
               "course": "course-v1:TestX+CRT-101+2015_1",
               "content_type": "Text",
               "content_groups": [
                  1756288587,
                  1104070958
               ],
               "id": "block-v1:TestX+CRT-101+2015_1+type@html+block@cb91b1d7ea2247c7a9b6992eb63b2c7e"
            }
         },
         {
            "_index": "courseware_index",
            "_type": "courseware_content",
            "_id": "block-v1:TestX+CRT-101+2015_1+type@html+block@461b0fae2dc24612828572e2ad2c58fc",
            "_score": 1,
            "_source": {
               "start_date": "2015-01-01T00:00:00+00:00",
               "content": {
                  "html_content": "This is content for cohort A",
                  "display_name": "Cohort A"
               },
               "course": "course-v1:TestX+CRT-101+2015_1",
               "content_type": "Text",
               "content_groups": [
                  1756288587,
                  1104070958
               ],
               "id": "block-v1:TestX+CRT-101+2015_1+type@html+block@461b0fae2dc24612828572e2ad2c58fc"
            }
         }
```

This test was simply one vertical with 2 html controls one for content group a and one for content group b - but they each get both because `groups_usage_info` is populated only with the vertical information:
```
{u'block-v1:TestX+CRT-101+2015_1+type@vertical+block@afd870b134a248dfac4df42b8b167265': [1756288587, 1104070958]}
```

Seems like a little additional refinement in building `groups_usage_info` is needed, then the code should work well.

dino-cikatic commented:
I've did several changes to code. First off, I've changed fetch_group_usage to return combined usage array, units and problems combined. After that I've removed inheriting permissions from parent since it's not needed (and is actually incorrect). Now children return their usage and if there is even one child open for public usage parent is not group restricted. 
Please take a look over changes.

dino-cikatic commented:
jenkins test this please

martynjames commented on lms/lib/courseware_search/lms_filter_generator.py:
```diff
@@ -3,13 +3,37 @@
+                pass
+            if partition_group:
+                partition_group = partition_group.group_id
+            filter_dictionary['content_groups'] = unicode(partition_group)
```
If partition_group is `None`, then the filter_dictionary gets an entry for `{'content_groups' : u'None'}` which is not going to do the right thing.

martynjames commented on lms/lib/courseware_search/lms_filter_generator.py:
```diff
@@ -3,13 +3,37 @@
+            except CourseUserGroupPartitionGroup.DoesNotExist:
+                pass
+            if partition_group:
+                partition_group = partition_group.group_id
```
@dino-cikatic Strange reassignment of partition_group to its child member. I'd prefer that we do something like this:
`filter_dictionary['content_groups'] = unicode(partition_group.group_id) if partition_group else None`

then add support within edx-search to allow a filter_dictionary specification to be `None`, which would allow us to always filter by a field (like `content_groups`) even when there is no value for that field.

I have taken the liberty of implementing this change in this PR: https://github.com/edx/edx-search/pull/13. Please review and one that is merged we'll want to update the has for edx-search within github.txt

dsego commented:
Should this PR be reopened against master?

martynjames commented:
We discussed on HipChat - in order to preserve the change history we'll keep the review against feature/dashboard-search. Because that branch has already merged, we can merge it again up to master after this review is complete

martynjames commented:
@cahrens 
After a couple of updates, this is now ready for T&L review, please.

_Description:_
This PR represents the work necessary to enable courseware search to support searching of cohorted content. The first instance of courseware search did not support cohorted content; this represents the changes necessary to deliver such. (see https://openedx.atlassian.net/browse/SOL-495)

_Sandbox:_
http://pr7640.m.sandbox.edx.org

andy-armstrong commented:
@dino-cikatic @dsego @martynjames @explorerleslie @marcotuts I'm performing the T&L review of this feature. I tested on the sandbox and everything looks great.

The one issue I came across was with preview when searching as a staff user. In this situation, I can choose to view the course as staff or as if I am a user with certain content group access. However, no matter the setting, I am never seeing cohorted courseware results. I would expect to see all content when previewing the course as staff, and I would expect when previewing as in a particular content group to see the appropriate content for that group.

To reproduce this:
1. Log in as staff@example.com
2. Navigate to this page in my test course:

http://pr7640.m.sandbox.edx.org/courses/course-v1:AndyA+AA101+1/courseware/c47fa8da17e4425f95f7f09ba918560c/56d65573ad09478f963c0db1becec1ef/

You should see four pieces of content when viewing the course as staff, but if you search for 'welcome' you will only get one result. Similarly, if you search for 'cat' you don't see any results.

3. Switch to "View this course as: Student in Cat Content"

You should now only see three pieces of content (the dog-only content is not shown). When I search for 'welcome' I also get only one result, and no results for 'cat'.

I imagine that the code is not obeying the preview masquerade, so I'll suggest where this change might be made. I'm also going to perform my high level code review now too.

andy-armstrong commented on common/test/acceptance/tests/lms/test_lms_cohorted_courseware_search.py:
```diff
@@ -0,0 +1,244 @@
+        self._auto_auth(self.USERNAME, self.EMAIL, False)
+        self.courseware_search_page.visit()
+
+    def test_cohorted_search_user_a_a_content(self):
```
Add a test to verify that staff users see all content by default. Then add a second test which verifies that they see only the appropriate content when previewing as belonging to a particular content group.

andy-armstrong commented on lms/lib/courseware_search/lms_filter_generator.py:
```diff
@@ -3,13 +3,35 @@
+                course_key = CourseKey.from_string(kwargs['course_id'])
+            except InvalidKeyError:
+                course_key = SlashSeparatedCourseKey.from_deprecated_string(kwargs['course_id'])
+            cohort = get_cohort(kwargs['user'], course_key, assign=False)
```
This code directly bypasses the cohort scheme object, and hence doesn't obey the preview masquerade logic. Instead, you should mimic the logic in ``lms/djangoapps/courseware/access.py``. That is looking up the group for each partition, but in your case you only need to do it for the one. Your code should be something like:

    cohorted_user_partition = get_cohorted_user_partition(course_key)
    partition_group = partition.scheme.get_group_for_user(
            course_key,
            user,
            cohorted_user_partition,
        )

Let me know if this doesn't work or if you have any questions.

andy-armstrong replied:
I realized on my walk home that my proposed solution doesn't cover the case of a staff user seeing all content. In the case that the user is staff and is not masquerading as a student, then there should be no content_groups filter added. I think you need to call ``has_access(user, 'staff', course_key)`` to determine if the user has staff access to the course.

andy-armstrong commented:
@dino-cikatic @dsego @martynjames - I'm done with my review. The code looks great, and the tests are impressively thorough. I've described how I would fix the bug I found, and then I'd add a couple of tests to prove that this behavior is now working. :+1: with those changes.

martynjames commented:
Thanks @andy-armstrong - I'd previously talked to @marcotuts about updating the staff views in a separate story, but you seem to have provided the solution right here! Thank you!

edx-webhook commented:
Thanks for the pull request, @dino-cikatic! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: http://openedx-webhooks.herokuapp.com/github/process_pr?number=7640&repo=edx%2Fedx-platform

edx-webhook commented:
Thanks for the pull request, @dino-cikatic! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: http://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=7640

martynjames commented:
Rebased against master in https://github.com/edx/edx-platform/pull/8043
